[torch.compile] Fuse RMSNorm with quant

ProExpertProg opened the pull request:
This PR enables fusing rms_norm and quant ops in the torch.compile backend. It adds all required infrastructure and new fused rms_norm_quant kernels. Only static FP8 quantization is supported in this PR, with more formats and datatypes to be added later to keep this PR as short as possible.

To enable fusion, `TORCH_COMPILE_LEVEL` needs to be at least 3, and the `RMSNorm` custom op needs to be enabled (by setting `TORCH_CUSTOM_OPS` either to `all` or `+rms_norm`). `TORCH_ENABLE_FUSION` needs to be set as well (on by default).

This PR gives roughly 2% end-to-end speedups: TODO detailed numbers.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

youkaichao committed:
adapt for dynamo

youkaichao committed:
fix tpu

youkaichao committed:
add backend

youkaichao committed:
add use_custom_dispatcher

youkaichao committed:
update wrapper

youkaichao committed:
update envs

youkaichao committed:
update custom op

youkaichao committed:
support llama

youkaichao committed:
update plugins

youkaichao committed:
update model runner

youkaichao committed:
add support

youkaichao committed:
add files

youkaichao committed:
fix not use_custom_dispatcher

ProExpertProg committed:
RMSNorm fusion working!

ProExpertProg committed:
fused with bug

youkaichao committed:
Merge branch 'main' into compile_integration

youkaichao committed:
do not test inductor

youkaichao committed:
add compile context

youkaichao committed:
remove model reference

youkaichao committed:
lint

ProExpertProg committed:
Use pattern matcher to match, replace manually, giving correct output

youkaichao committed:
change levels

youkaichao committed:
Merge branch 'main' into compile_integration

youkaichao committed:
add levels

youkaichao committed:
use const

youkaichao committed:
use const

youkaichao committed:
use const

youkaichao committed:
use const

youkaichao committed:
use const

youkaichao committed:
use const

youkaichao committed:
use const

youkaichao committed:
use const

youkaichao committed:
use const

youkaichao committed:
use const

youkaichao committed:
use const

youkaichao committed:
restore

youkaichao committed:
use const

youkaichao committed:
use const

youkaichao committed:
error on inductor for tpu

youkaichao committed:
fix llava

youkaichao committed:
restore tpu

youkaichao committed:
Merge branch 'main' into compile_integration

youkaichao committed:
adjust for tpu

youkaichao committed:
fix env var

youkaichao committed:
fix calling

youkaichao committed:
revert tpu

youkaichao committed:
revert utils

youkaichao committed:
fix typo

ProExpertProg committed:
add quant_layernorm kernel (not modified yet)

ProExpertProg committed:
fixes

ProExpertProg committed:
out -> result for fp8 quant ops

ProExpertProg committed:
change int8 to fp8

youkaichao committed:
add typing

youkaichao committed:
move DYNAMO_AS_IS to model runner level

youkaichao committed:
fix default context

youkaichao committed:
use eager for DYNAMO_AS_IS by default

youkaichao committed:
update tests

youkaichao committed:
update tests

youkaichao committed:
llava uses fullgraph=false

ProExpertProg committed:
Added layernorm_quant kernels for static fp8 quant, including tests. Fusion seems to work

ProExpertProg committed:
fix_functionalization for layernorm_quant kernels

ProExpertProg committed:
env var for disabling fusion

ProExpertProg committed:
Fix for fusion assert

youkaichao committed:
Merge branch 'main' into compile_integration

youkaichao committed:
Merge branch 'main' into compile_integration

youkaichao committed:
disable tests first

ProExpertProg committed:
Clean up fusion.py

youkaichao committed:
Merge branch 'main' into compile_integration

youkaichao committed:
add supports_dynamo in the decorator

ProExpertProg committed:
Merge branch 'compile_integration' into luka/rms-norm-fusion-kaichao

ProExpertProg committed:
fix example_inputs dtype

ProExpertProg committed:
Merge remote-tracking branch 'upstream/main' into luka/rms-norm-fusion

# Conflicts:
#	vllm/compilation/backends.py
#	vllm/compilation/decorators.py
#	vllm/envs.py
#	vllm/model_executor/models/gemma2.py
#	vllm/model_executor/models/llama.py

bnellnm commented on vllm/compilation/fusion.py:
```diff
@@ -0,0 +1,167 @@
+        find_getitem(quant_node, 1).replace_all_uses_with(result_node_new)
+
+    # Finally, remove matched nodes
+    graph.eliminate_dead_code()
```
Have you verified this gets rid of all the old crud code?

ProExpertProg replied:
Yep, that's what the assert is for, if you're referring to the pattern we replaced.

ProExpertProg committed:
extract common type conversion stuff to .cuh file

ProExpertProg committed:
Merge remote-tracking branch 'refs/remotes/upstream/main' into luka/rms-norm-fusion

# Conflicts:
#	vllm/envs.py

ProExpertProg committed:
format

bnellnm commented on vllm/compilation/fusion.py:
```diff
@@ -0,0 +1,224 @@
+    for match in matches:
+        nodes = list(graph.nodes)
+        # TODO this is an expensive check
+        if not all(node in nodes for node in match.nodes):
+            raise ValueError(
+                f"Broken match: not all nodes in graph: "
+                f"{[node for node in match.nodes if node not in nodes]}")
```
This check seems like overkill/unnecessary.

ProExpertProg committed:
PR comments

Remove expensive check

bnellnm commented on csrc/type_convert.cuh:
```diff
@@ -0,0 +1,167 @@
+#ifndef USE_ROCM
+  #include <cuda_bf16.h>
+  #include <cuda_fp16.h>
+//  #include <cub/util_type.cuh>
```
nit: unused include

bnellnm commented on vllm/compilation/backends.py:
```diff
@@ -93,28 +94,75 @@ def fix_functionalization(graph: fx.Graph):
+                result = kwargs['result']
+                residual = kwargs['residual']
+
+                # Create a new call to torch.ops._C.rotary_embedding.default
```
nit: update comment

bnellnm commented on vllm/compilation/backends.py:
```diff
@@ -93,28 +94,75 @@ def fix_functionalization(graph: fx.Graph):
-                weight = kwargs['weight']
-                epsilon = kwargs['epsilon']
+                replace_node = kwargs['result']
                 # Create a new call to torch.ops._C.rotary_embedding.default
```
ditto

bnellnm commented on vllm/compilation/backends.py:
```diff
@@ -93,28 +94,75 @@ def fix_functionalization(graph: fx.Graph):
+                kwargs = node.kwargs
+
+                replace_node = kwargs['result']
+                # Create a new call to torch.ops._C.rotary_embedding.default
```
Looks like a lot of these comments are cut/paste and need to be updated.

bnellnm commented on vllm/compilation/backends.py:
```diff
@@ -167,8 +223,8 @@ def wrap_inductor(graph, example_inputs, additional_inductor_config):
             "post_grad_custom_post_pass is already set in the config. "
-            "Overwriting it with the fix_functionalization")
-    current_config['post_grad_custom_post_pass'] = fix_functionalization
+            "Overwriting it with the post_grad_post_passes")  # TODO combine
```
address TODO?

bnellnm commented on vllm/compilation/fusion.py:
```diff
@@ -0,0 +1,219 @@
+        empty_bf16(1, 5),
+        torch.empty(1, 1, device="cuda", dtype=torch.float32)
+    ]
+    register_replacement(rms_pattern_residual_static,
```
Can you add a comment somewhere that explains why we need to use the callback and do manual replacement?  A link to the issue would be helpful too.
https://github.com/pytorch/pytorch/issues/137280

bnellnm commented on vllm/compilation/fusion.py:
```diff
@@ -0,0 +1,219 @@
+def process_matches(matches, graph: torch.fx.Graph):
+    for match in matches:
+        nodes = list(graph.nodes)
+        last_node_in_match = max(match.nodes, key=lambda x: nodes.index(x))
```
wouldn't `nodes[-1]` give the same result?

ProExpertProg replied:
Yes, I believe so, from inspecting values in the debugger. I can look again though

bnellnm replied:
Ok, I'll buy that.  The comments in the code don't give any explicit invariants on the Match nodes so I guess we have to account for that.   I wonder if there's a more efficient way to do this though (I'm fine with figuring it out later).

ProExpertProg replied:
I think we could just do a single pass over the list of nodes instead of finding each one. This was just the first pass and inspired by some of the code in the pattern matcher IIRC.

ProExpertProg replied:
Just checked in the debugger, last node in graph is `getitem_1590`:
`match.nodes`: `[auto_functionalized_410, getitem_1587, auto_functionalized_411, getitem_1590, getitem_1588]`
Graph:
```python

     # File: /home/luka/neuralmagic-vllm/vllm/_custom_ops.py:213 in fused_add_rms_norm, code: torch.ops._C.fused_add_rms_norm(input, residual, weight, epsilon)
    auto_functionalized_410 = torch.ops.higher_order.auto_functionalized(torch.ops._C.fused_add_rms_norm.default, input = getitem_1585, residual = getitem_1551, weight = arg479_1, epsilon = 1e-05);  getitem_1585 = getitem_1551 = arg479_1 = None
    getitem_1587: "f16[s0, 4096]" = auto_functionalized_410[1]
    getitem_1588: "f16[s0, 4096]" = auto_functionalized_410[2];  auto_functionalized_410 = None
    
     # File: /home/luka/neuralmagic-vllm/vllm/_custom_ops.py:675 in scaled_fp8_quant, code: output = torch.empty(shape, device=input.device, dtype=out_dtype)
    empty_284: "f8e4m3fn[s0, 4096]" = torch.ops.aten.empty.memory_format([arg0_1, 4096], dtype = torch.float8_e4m3fn, device = device(type='cuda', index=0), pin_memory = False)
    
     # File: /home/luka/neuralmagic-vllm/vllm/_custom_ops.py:690 in scaled_fp8_quant, code: torch.ops._C.static_scaled_fp8_quant(output, input, scale)
    auto_functionalized_411 = torch.ops.higher_order.auto_functionalized(torch.ops._C.static_scaled_fp8_quant.default, result = empty_284, input = getitem_1587, scale = arg482_1);  empty_284 = getitem_1587 = None
    getitem_1590: "f8e4m3fn[s0, 4096]" = auto_functionalized_411[1];  auto_functionalized_411 = None
```

bnellnm commented on vllm/compilation/fusion.py:
```diff
@@ -0,0 +1,219 @@
+    patterns, matches = get_patterns()
+
+    def dump_graph(graph: torch.fx.Graph, stage: str):
+        if stage in envs.VLLM_TORCH_COMPILE_FUSION_DUMP:
+            logger.info("Printing graph to %s", f"{stage}.py")
+            with open(f"{stage}.py", "w") as f:
+                print(graph.python_code(root_module="self", verbose=True).src,
+                      file=f)
```
You might want to add the rank to the filename to avoid problems when tp_size > 1 (or only add the rank to the name when tp_size > 1).

ProExpertProg replied:
How do I find out the rank? Should I just read envs.LOCAL_RANK?

bnellnm replied:
you can use `get_tensor_model_parallel_rank()`.

tlrmchlsmth commented on csrc/type_convert.cuh:
Nice refactor

tlrmchlsmth commented on csrc/type_convert.cuh:
```diff
@@ -0,0 +1,167 @@
+#pragma once
+
+#include <torch/all.h>
```
I think including `torch/all.h` is probably overkill and pulls in a bunch of stuff we don't need.

Could you try including `torch/cuda.h` and maybe add `torch/torch.h` if that's not enough?

tlrmchlsmth commented on csrc/quantization/fp8/common.cuh:
```diff
@@ -0,0 +1,185 @@
+#pragma once
+
+// TODO(luka) remove unnecessary includes
```
Reminder to address TODO

tlrmchlsmth commented on csrc/layernorm_quant_kernels.cu:
Could you add a high-level comment at the top of this file on what these kernels are for?

tlrmchlsmth commented on vllm/compilation/fusion.py:
Does it make sense to add a unit test for the fusion pass?

SageMoore replied:
We don't necessarily need it for this diff, but I think it would be good to get in the habit of writing unit tests for the passes. Even just running a naive example through torch compile and confirming the fused kernel is substituted as expected and that it gets the same answer as the unfused version would be sufficient. That would give decent coverage for the multi output rewrite stuff along with the actual pattern matching.

A simple utility that searches the compiled graph for the expected kernel shouldn't be too bad. Again, I'm totally fine with this being a follow on, but I do think we should add some test infra here.

ProExpertProg replied:
Will add in follow-up PR

SageMoore commented on vllm/compilation/fusion.py:
```diff
@@ -0,0 +1,219 @@
+                print(graph.python_code(root_module="self", verbose=True).src,
+                      file=f)
+
+    def fusion_pass(graph: torch.fx.Graph):
```
Instead of capturing, I think it would be cool to refactor this a bit so that `fusion_pass` is a callable that explicitly owns `matches` and `patterns`.

SageMoore commented on vllm/compilation/fusion.py:
```diff
@@ -0,0 +1,219 @@
+        rms_val = rms_node.meta["val"]
+        quant_val = quant_node.meta["val"]
+        fused_node.meta["val"] = (None, quant_val[1], rms_val[1], rms_val[2])
+
```
Something like: `# Replace the existing outputs with the new getitem nodes. The eliminate_dead_code call below will delete these now orphaned nodes from the graph` ?

SageMoore commented on vllm/compilation/fusion.py:
```diff
@@ -0,0 +1,219 @@
+    for match in matches:
+        nodes = list(graph.nodes)
+        last_node_in_match = max(match.nodes, key=lambda x: nodes.index(x))
+        with graph.inserting_after(last_node_in_match):
```
This section feels like it needs a comment. Something to describe why we are adding these getitem nodes along with the general tuple structure of the auto_functionalized output.

SageMoore reviewed (commented):
Minimal comments. Otherwise looks good. Nice work.

ProExpertProg committed:
refactored fusion pass into a class

ProExpertProg committed:
PR comments: backends.py
- fix_func pass fixed
- support additional passes through config

ProExpertProg committed:
Merge branch 'main' into luka/rms-norm-fusion

ProExpertProg committed:
Fix node bug in find_getitem

ProExpertProg committed:
PR comments:
- add TP rank to dump filename
- improve search for last node in match
- massively improve comments

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
PR comments:
- remove unnecessary includes
- add comment in layernorm quant

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
yapf fix

Signed-off-by: luka <luka@neuralmagic.com>

mgoin reviewed (approved):
Excellent work

ProExpertProg committed:
Merge remote-tracking branch 'upstream/main' into luka/rms-norm-fusion

ProExpertProg committed:
Unit test for rmsnorm-quant fusion

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Skip on non-CUDA

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Merge remote-tracking branch 'upstream/main' into luka/rms-norm-fusion

ProExpertProg committed:
Fix FP8 HIP type in common.cuh

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Merge remote-tracking branch 'upstream/main' into luka/rms-norm-fusion

ProExpertProg committed:
Fix seed_everything

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Merge remote-tracking branch 'upstream/main' into luka/rms-norm-fusion

# Conflicts:
#	vllm/compilation/backends.py

ProExpertProg committed:
Add support for passes to VllmBackend, add fusion back in

Signed-off-by: luka <luka@neuralmagic.com>

bnellnm reviewed (approved):
LGTM!

youkaichao commented on vllm/compilation/backends.py:
```diff
@@ -261,14 +316,27 @@ class VllmBackend:
         # do anything here
 
+    def add_passes_to_config(self):
+        config = self.compilation_configs.inductor_compile_config
```
why do we move it from `inductor_passes` to `inductor_compile_config`?

I separate them out, because we need to convert function qualname to function object in passes, while the rest configs are taken as-is.

ProExpertProg replied:
We don't, I left that part unchanged. This is just additional built-in passes that go to `ppst_grad_custom_post_pass`.

youkaichao commented on vllm/envs.py:
```diff
@@ -225,6 +227,16 @@ def get_default_config_root():
     "VLLM_CUSTOM_OPS":
     lambda: os.environ.get("VLLM_CUSTOM_OPS", "").replace(" ", "").split(","),
+
+    # Internal flag to enable fusion in torch.compile (default on)
+    "VLLM_TORCH_COMPILE_FUSION":
+    lambda: bool(os.environ.get("VLLM_TORCH_COMPILE_FUSION", "1") != "0"),
```
we can add a new field in the config

ProExpertProg replied:
Done

youkaichao commented on vllm/envs.py:
```diff
@@ -225,6 +227,16 @@ def get_default_config_root():
+    "VLLM_TORCH_COMPILE_FUSION":
+    lambda: bool(os.environ.get("VLLM_TORCH_COMPILE_FUSION", "1") != "0"),
+
+    # Internal flag for dumping the model graph at different stages of
+    # custom pass compilation
+    "VLLM_TORCH_COMPILE_DUMP":
+    lambda: list(os.environ.get("VLLM_TORCH_COMPILE_DUMP", "").split(",")),
```
we can add a new field in config.

ProExpertProg replied:
Done

ProExpertProg committed:
Merge remote-tracking branch 'upstream/main' into luka/rms-norm-fusion

ProExpertProg committed:
PR comments:
 - move extra env variables to config
 - fix pass adding logic

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Fix fusion pass init in test

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Fusion test: use apply_fp8_linear

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Add redundant reshapes removal pass.

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Fix graph dumping when TP not initialized

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Reshape add edge-cases

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Singleton pattern matcher for fusion pass

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Add print

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
singleton fusion pass

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Merge remote-tracking branch 'upstream/main' into luka/rms-norm-fusion

Signed-off-by: luka <luka@neuralmagic.com>

# Conflicts:
#	vllm/compilation/backends.py

ProExpertProg committed:
format

Signed-off-by: luka <luka@neuralmagic.com>

youkaichao commented on vllm/compilation/fusion.py:
```diff
@@ -0,0 +1,291 @@
+                   for node in match.nodes)
+
+    def __call__(self, graph: torch.fx.Graph):
+        self.dump_graph(graph, "before_fusion")
```
we can use pytorch's built-in `lazy_format_graph_code`, see https://github.com/vllm-project/vllm/blob/f6778620a95baf925eb54694ab4666524d0d8584/vllm/compilation/backends.py#L365

ProExpertProg replied:
For me, I like the graph going to a file because I can keep around multiple versions, compare them, and navigate more easily. If it's all printed to the console, I'd most likely need to copy each of the graphs to a file manually.

But maybe we can make it configurable so we could do both?

youkaichao commented on vllm/compilation/inductor_pass.py:
```diff
@@ -0,0 +1,38 @@
+
+            logger.info("Printing graph to %s", filepath)
+            with open(filepath, "w") as f:
+                src = graph.python_code(root_module="self", verbose=True).src
```
thinks is quite similar to what i did in https://github.com/thuml/depyf . I did quite a lot hacking to make sure the dumped code is readable and parsable by IDEs.

I think the logging infra can be designed together with compilation cache infra.

youkaichao reviewed (approved):
LGTM, I left some comments, but we don't need to address them in this PR. thanks for your hard work!

tlrmchlsmth commented on tests/compile/test_fusion.py:
TYVM for adding this test
